drivers/bluetooth/hci/hci_da1469x.c

```diff
@@ -0,0 +1,495 @@
+
+			LOG_DBG("Calling bt_recv(%p)", buf);
+			bt_recv(buf);
+
+			/* Give other threads a chance to run if the ISR
+			 * is receiving data so fast that rx.fifo never
+			 * or very rarely goes empty.
+			 */
+			k_yield();
+
+			irq_disable(CMAC2SYS_IRQn);
+			buf = net_buf_get(&rx.fifo, K_NO_WAIT);
+		} while (buf);
+	}
+}
+
+static size_t h4_discard(size_t len)
+{
+	uint8_t buf[33];
+	int err;
+
+	err = cmac_mbox_read(buf, MIN(len, sizeof(buf)));
+	if (unlikely(err < 0)) {
+		LOG_ERR("Unable to read from UART (err %d)", err);
+		return 0;
+	}
+
+	return err;
+}
+
+static inline void read_payload(void)
+{
+	struct net_buf *buf;
+	int read;
+
+	if (!rx.buf) {
+		size_t buf_tailroom;
+
+		rx.buf = get_rx(K_NO_WAIT);
+		if (!rx.buf) {
+			if (rx.discardable) {
+				LOG_WRN("Discarding event 0x%02x", rx.evt.evt);
+				rx.discard = rx.remaining;
+				reset_rx();
+				return;
+			}
+
+			LOG_WRN("Failed to allocate, deferring to rx_thread");
+			irq_disable(CMAC2SYS_IRQn);
+			return;
+		}
+
+		LOG_DBG("Allocated rx.buf %p", rx.buf);
+
+		buf_tailroom = net_buf_tailroom(rx.buf);
+		if (buf_tailroom < rx.remaining) {
+			LOG_ERR("Not enough space in buffer %u/%zu", rx.remaining, buf_tailroom);
+			rx.discard = rx.remaining;
+			reset_rx();
+			return;
+		}
+
+		copy_hdr(rx.buf);
+	}
+
+	read = cmac_mbox_read(net_buf_tail(rx.buf), rx.remaining);
+	if (unlikely(read < 0)) {
+		LOG_ERR("Failed to read UART (err %d)", read);
+		return;
+	}
+
+	net_buf_add(rx.buf, read);
+	rx.remaining -= read;
+
+	LOG_DBG("got %d bytes, remaining %u", read, rx.remaining);
+	LOG_DBG("Payload (len %u): %s", rx.buf->len, bt_hex(rx.buf->data, rx.buf->len));
+
+	if (rx.remaining) {
+		return;
+	}
+
+	buf = rx.buf;
+	rx.buf = NULL;
+
+	if (rx.type == H4_EVT) {
+		bt_buf_set_type(buf, BT_BUF_EVT);
+	} else {
+		bt_buf_set_type(buf, BT_BUF_ACL_IN);
+	}
+
+	reset_rx();
+
+	LOG_DBG("Putting buf %p to rx fifo", buf);
+	net_buf_put(&rx.fifo, buf);
+}
+
+static inline void read_header(void)
+{
+	switch (rx.type) {
+	case H4_NONE:
+		h4_get_type();
+		return;
+	case H4_EVT:
+		get_evt_hdr();
+		break;
+	case H4_ACL:
+		get_acl_hdr();
+		break;
+	case H4_ISO:
+		if (IS_ENABLED(CONFIG_BT_ISO)) {
+			get_iso_hdr();
+			break;
+		}
+		__fallthrough;
+	default:
+		CODE_UNREACHABLE;
+		return;
+	}
+
+	if (rx.have_hdr && rx.buf) {
+		if (rx.remaining > net_buf_tailroom(rx.buf)) {
+			LOG_ERR("Not enough space in buffer");
+			rx.discard = rx.remaining;
+			reset_rx();
+		} else {
+			copy_hdr(rx.buf);
+		}
+	}
+}
+
+static inline void process_rx(void)
+{
+	LOG_DBG("remaining %u discard %u have_hdr %u rx.buf %p len %u", rx.remaining, rx.discard,
+		rx.have_hdr, rx.buf, rx.buf ? rx.buf->len : 0);
+
+	if (rx.discard) {
+		rx.discard -= h4_discard(rx.discard);
+		return;
+	}
+
+	if (rx.have_hdr) {
+		read_payload();
+	} else {
+		read_header();
+	}
+}
+
+void cmac_read_req(void)
+{
+	while (cmac_mbox_has_data()) {
+		process_rx();
+	}
+}
+
+void cmac_rng_req(void)
+{
+	k_sem_give(&rng_sem);
+}
```

Comment:
Maybe add code comments explaining where the calls to these functions are coming from, since they're not in the main Zephyr tree.